### ISSUE TITLE
Admin routes for custom_api tasks

### ISSUE THREAD

**Issue by Marchand-Nicolas** [starknet-id/quest_server#266]
## Description

Implement the necessary admin routes to create and edit a new `custom_api` task type.

## PROPOSED TODO

- [ ] Create a `custom_api` folder in `src\endpoints\admin` containing a `create_custom_api` and an `update_custom_api` route (such as in `src\endpoints\admin\balance`)
- [ ] It should take a `api_url` (**string**) field and `regex` (**string**) field as arguments in the body (among the other default fields such as the task name), and store it in the task document in the db (as we are doing for other task types, such as in `src\endpoints\admin\balance`.)
- [ ] Add the `api_url` (**string**) and `regex` (**string**) optional fields to the `QuestTaskDocument` type in `src\models.rs` if they are not already present.
- [ ] The new document should look something like that:

```rs
    let new_document = QuestTaskDocument {
        name: body.name.clone(),
        desc: body.desc.clone(),
        verify_redirect: None,
        href: body.href.clone(),
        total_amount: None,
        quest_id: body.quest_id,
        id: next_id,
        cta: body.cta.clone(),
        verify_endpoint: "quests/verify_custom_api".to_string(),
        verify_endpoint_type: "default".to_string(),
        task_type: Some("custom_api".to_string()),
        discord_guild_id: None,
        quiz_name: None,
        contracts: None,
        api_url: Some(body.api_url),
        regex: Some(body.regex),
    };
```

**Comment by NueloSE** [comment]
hi @Marchand-Nicolas kindly assign this task to me. it would be nice if i can make it my first contribution to this repo

**Comment by mubarak23** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/starknet-quest).

### My background and how it can be leveraged
I am a Experience Cairo smart contract developer with experience working on projects such as Just Art Peace, Dojo, Kart, TBA, and Shinigami. Before transitioning to Cairo development, I was a backend developer specializing in Rust.



My recent work with cairo starknet
- https://github.com/keep-starknet-strange/raito/pull/128
- https://github.com/keep-starknet-strange/raito/pull/143
- https://github.com/horuslabsio/TBA/pull/45
- https://github.com/horuslabsio/karst-core/pull/60
- https://github.com/keep-starknet-strange/art-peace/pull/121
- https://github.com/keep-starknet-strange/zkramp/pull/33
- https://github.com/keep-starknet-strange/art-peace/pull/32
- https://github.com/horuslabsio/TBA/pull/53



My recent work with rust
- https://github.com/mubarak23/axum-user-auth
- https://github.com/antoyo/rustc_codegen_gcc/pull/1
- https://github.com/antoyo/rustc_codegen_gcc/pull/5
- https://github.com/fedimint/fedimint/pull/5994
- https://github.com/mubarak23/soroban-bank



### How I plan on tackling this issue
I have read the description

This issue, am going to create two endpoints
- CREATE CUSTOM TASK
   - collect the request body payload
   - validate the request body payload
   - make sure the payload contain api_url and regex
   - save the payload in the new db model for custom api TASK
   - if save fail, return the right error response
   - if save is successful, return successful response

- EDIT CUSTOM API TASK
  - collect the id in the route param
  - collect the payload
  - fetch the custom api task, base on the ID
  - if the custom API TASK does not exist, resturn an error response 
  - if the custom API TASK exist, validate the update payload and update the record
  - return error if update fail
  - return successful response if update operation was successful.

ETH: 16HRS

**Comment by PabloVillaplana** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/starknet-quest).

### My background and how it can be leveraged
I am a Full Stack developer with 6 years of experience in the industry, I am super excited to be able to contribute to Web 3 projects, this to continue paving my way as an Open Source Contributor. I am from Dojo Coding Community

### How I plan on tackling this issue
I would like to contribute again to Starknet Quest with this issue, I have to do a little research, but I would like to be able to contribute to this issue in Cairo, since it would help me a lot to be able to raise my level in Starknet.

**Comment by Ugo-X** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/starknet-quest).

### My background and how it can be leveraged
 

I am a Full Stack blockchain Developer with expertise in Next.js, Nest.js, TypeScript, JavaScript, React, Node.js, Three.js, Solidity and rust. My journey with OnlyDust hackathons began at Edition 1, and I've since made 69 contributions across 14 projects. With my extensive experience on the OnlyDust platform (profile:  https://app.onlydust.com/u/Ugo-X), I've honed my skills in delivering quality solutions under pressure. 

I bring a unique blend of technical proficiency and user-focused design to every project, whether it's crafting immersive 3D experiences or developing smart contracts. My track record shows I can adapt quickly and contribute effectively to diverse challenges. 

As we surf through Edition 8, I'm excited to leverage my skills and hackathon experience to push the boundaries of blockchain development. I'm confident in my ability to tackle new challenges and drive innovation in this space.

### How I plan on tackling this issue
I will approach this issue in the following manner: 

  

1. Create a new folder `custom_api` in `src\endpoints\admin`. 

2. Implement `create_custom_api.rs`: 

   a. Define a request body struct with fields for task details, `api_url`, and `regex`. 

   b. Validate input data, ensuring `api_url` is a valid URL and `regex` is a valid regular expression. 

   c. Generate a new task ID. 

   d. Create a new `QuestTaskDocument` with "custom_api" task type and provided data. 

   e. Insert the document into the database. 

   f. Return a success response with the created task details. 

3. Implement `update_custom_api.rs`: 

   a. Define a request body struct similar to create, but with an `id` field. 

   b. Retrieve the existing task document by ID. 

   c. Update the document with new data, preserving unchanged fields. 

   d. Save the updated document to the database. 

   e. Return a success response with the updated task details. 

4. Add error handling for database operations and input validation. 

5. Update the admin API router to include these new routes.

**Comment by Dprof-in-tech** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/starknet-quest).

### My background and how it can be leveraged
Hello, i am Dprof-in-tech, an experienced Full Stack Blockchain Developer and  I am excited to contribute my skills to the {project name} project in this ODHACK 8. With a strong background in Next.js, TypeScript, JavaScript, React, Node.js, Rust and Cairo, I've honed my technical skills across the blockchain development landscape.
My journey with OnlyDust began at Edition 2, and I've since made 34 contributions across 11 projects. This extensive experience on the platform has allowed me to develop a keen understanding of delivering high-quality solutions under tight deadlines.
Due to my web2 experience, i am skilled in writing unit tests and have used Jest and Vitest to do this. I would love to contribute to this project by solving this issue.
Below is a link to my OnlyDust public profile.
https://app.onlydust.com/u/Dprof-in-tech


### How I plan on tackling this issue
To solve this issue, i would :



1. I’ll create a `custom_api` folder under the `admin` directory, with routes for `create_custom_api` and `update_custom_api`.

2. The new task document will be updated to include the `api_url` and `regex` fields alongside existing fields like `name`, `description`, etc.

3. I’ll also add these fields to the database schema, ensuring we store them correctly.

4. The `create` route will allow creating tasks with the `custom_api` type, while the `update` route will handle modifying existing tasks.

5. Finally, I’ll implement the necessary logic to interact with the database for task creation and updates, making sure the new fields are properly handled.

Once this is done, I’ll run tests to confirm everything is working as expected.

I’ll be sure to follow the guidelines as given and work to tender my solution as soon as possible.

**Comment by GoSTEAN** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/starknet-quest).

### My background and how it can be leveraged
I have experience in backend development, specifically in creating and managing API routes using frameworks like Node.js, Rust, and TypeScript. I’ve worked on similar task management systems, where I handled database integration and ensured routes were properly set up to handle custom data. My familiarity with task-based APIs and database schemas will help me effectively implement and extend the custom API task type as required.



### How I plan on tackling this issue
I would begin by creating the necessary folder and routes for `create_custom_api` and `update_custom_api`. Then, I’d add the `api_url` and `regex` fields to the request body and ensure they are correctly stored in the task document within the database. I’d update the `QuestTaskDocument` type to include these optional fields if they aren't already present. Finally, I would test the endpoints to ensure they function correctly with the new fields and integrate seamlessly with the existing API structure.

**Comment by GideonBature** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/starknet-quest).

### My background and how it can be leveraged
I am a Software Engineer with experience in building a complete backend in Rust using RESTful API. I also have experience working with JavaScript and TypeScript.

### How I plan on tackling this issue
First of all I will look at the code base and understand the structure, next I will look at other tasks type that are already present and implemented in the codebase to see the pattern used and know which kind of database that is used, this will guide me to know how I am going to structure my own code and implementation. Finally, I will add the `api_url` and the `regex`string as fields in the already existing `QuestTaskDocument` struct.

Adding them as an option means they can either be present as value or None.

**Comment by GideonBature** [comment]
Thank you @Marcmehand-Nicolas...

**Comment by Marchand-Nicolas** [comment]
Done here https://github.com/lfglabs-dev/api.starknet.quest/pull/272
